libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/ptx/ptx_helper_functions.h

```diff
@@ -26,37 +27,82 @@ _LIBCUDACXX_BEGIN_NAMESPACE_CUDA_PTX
 
 inline _LIBCUDACXX_DEVICE _CUDA_VSTD::uint32_t __as_ptr_smem(const void* __ptr)
 {
-  // Consider adding debug asserts here.
-  return static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(__ptr));
+  NV_DISPATCH_TARGET(
+    NV_IS_DEVICE, (
+      // Consider adding debug asserts here.
+      return static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(__ptr));
+    ),
+    NV_ANY_TARGET, (
+      // UNREACHABLE
+      _LIBCUDACXX_UNREACHABLE();
```

Comment:
suggestion: as we discussed offline, we might want the same approach as in ptx wrappers here (and in the remaining places) with linkage error. If there's value in merging this PR sooner, let's file an issue and address this later.

Reply:
It occurs to me that it would be a breaking-ish change if we make what was previously a runtime error into a compile time error. Given we have plans to always make things like this eventually be a compile time error, I think we should opt for always making them link time errors today.

Reply:
I've submitted the changes that *do* modify them into a breaking change.